src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md

```diff
@@ -0,0 +1,74 @@
+# `thumbv7em-none-eabi` and `thumbv7em-none-eabihf`
+
+**Tier: 2**
+
+Bare-metal target for CPUs in the [Armv7E-M] architecture family, supporting a
+subset of the [T32 ISA][t32-isa].
+
+Processors in this family include the:
+
+* [Arm Cortex-M4][cortex-m4] and Arm Cortex-M4F
+* [Arm Cortex-M7][cortex-m7] and Arm Cortex-M7F
+
+See [`arm-none-eabi`](arm-none-eabi.md) for information applicable to all
+`arm-none-eabi` targets, in particular the difference between the `eabi` and
+`eabihf` ABI.
+
+[t32-isa]: https://developer.arm.com/Architectures/T32%20Instruction%20Set%20Architecture
+[Armv7E-M]: https://developer.arm.com/documentation/ddi0403/latest/
+[cortex-m4]: https://developer.arm.com/Processors/Cortex-M4
+[cortex-m7]: https://developer.arm.com/Processors/Cortex-M7
+
+## Target maintainers
+
+* [Rust Embedded Devices Working Group Cortex-M
+  Team](https://github.com/rust-embedded), `cortex-m@teams.rust-embedded.org`
+
+## Target CPU and Target Feature options
+
+See [the bare-metal Arm
+docs](arm-none-eabi.md#target-cpu-and-target-feature-options) for details on how
+to use these flags.
+
+### Table of supported CPUs for `thumbv7em-none-eabi`
+
+| CPU        | FPU | DSP | Target CPU  | Target Features |
+| ---------- | --- | --- | ----------- | --------------- |
+| Any        | No  | Yes | None        | None            |
+| Cortex-M4  | No  | Yes | `cortex-m4` | `+soft-float`   |
+| Cortex-M4F | SP  | Yes | `cortex-m4` | None            |
+| Cortex-M7  | No  | Yes | `cortex-m7` | `+soft-float`   |
```

Comment:
FWIW this is a problematic piece of advice to give, since this target feature is also implicated in the soundness issue tracked at https://github.com/rust-lang/rust/issues/116344.

Reply:
Although as [we discussed extensively on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.60-Ctarget-feature.3D.2Bsoft-float.60.20considered.20harmful) currently we know of no other way of stopping `target-cpu=cortex-m7` (or similar) from emitting FPU instructions like `vadd.f32`, yet it's reasonable to want the Cortex-M7 specific instruction scheduling optimisations even if your Cortex-M7 doesn't have an FPU.

This feature is totally unsound on `*-unknown-none-eabihf`, but it's both sound and useful on `*-unknown-none-eabi`. The 'target-feature' checker can currently only look at the target-architecture, and not the full. target triple, so it's unclear where we go from here.

Reply:
The target-feature checker is still WIP, and will be improved to handle this, somehow.

But the docs maybe should call out that this same flag is unsound to use on other targets (until rustc reaches the point where it is able to tell users about this directly).